File: test/java/sqlrtest.java
// Copyright (c) David Muse
// See the file COPYING for more information.

import com.firstworks.sqlrelay.*;

class sqlrtest {

	protected static int status=0;

	protected static String success="\u001B[32msuccess\u001B[0m";
	protected static String failure="\u001B[31mfailure\u001B[0m";
	protected static String alltestssucceeded="\n\u001B[34mAll tests succeeded\u001B[0m";
	protected static String sometestsfailed="\n\u001B[33mSome tests failed\u001B[0m";

	protected static SQLRConnection con;
	protected static SQLRCursor cur;

	protected static void printErrors() {

		if (cur!=null) {
			String err=cur.errorMessage();
			if (err!=null) {
				System.out.println(err);
				return;
			}
		}
		if (con!=null) {
			String err=con.errorMessage();
			if (err!=null) {
				System.out.println(err);
				return;
			}
		}
	}

	protected static void assertEquals(String actual,
						String expected, int length) {

		if (expected==null) {
			if (actual==null) {
				System.out.print(success+" ");
				return;
			} else {
				System.out.println(failure);
				System.out.println(actual+"!="+expected);
				printErrors();
				status=1;
			}
			return;
		}

		if (actual.regionMatches(0,expected,0,length)) {
			System.out.print(success+" ");
		} else {
			System.out.println(failure);
			System.out.println(actual+"!="+expected);
			printErrors();
			status=1;
		}
	}

	protected static void assertEquals(String actual, String expected) {

		if (expected==null) {
			if (actual==null) {
				System.out.print(success+" ");
				return;
			} else {
				System.out.println(failure);
				System.out.println(actual+"!="+expected);
				printErrors();
				status=1;
			}
			return;
		}

		if (actual!=null && actual.equals(expected)) {
			System.out.print(success+" ");
		} else {
			System.out.println(failure);
			System.out.println(actual+"!="+expected);
			printErrors();
			status=1;
		}
	}

	protected static void assertEquals(byte[] actual,
						String expected, int length) {

		if (expected==null) {
			if (actual==null) {
				System.out.print(success+" ");
				return;
			} else {
				System.out.println(failure);
				printErrors();
				status=1;
			}
			return;
		}

		byte[]	successvalue=expected.getBytes();

		for (int index=0; index<length; index++) {
			if (actual[index]!=successvalue[index]) {
				System.out.println(failure);
				printErrors();
				status=1;
				return;
			}
		}

		System.out.print(success+" ");
	}

	protected static void assertEquals(byte[] actual,
						byte[] expected, int length) {

		if (expected==null) {
			if (actual==null) {
				System.out.print(success+" ");
				return;
			} else {
				System.out.println(failure);
				printErrors();
				status=1;
			}
			return;
		}

		for (int index=0; index<length; index++) {
			if (actual[index]!=expected[index]) {
				System.out.println(failure);
				printErrors();
				status=1;
				return;
			}
		}

		System.out.print(success+" ");
	}

	protected static void assertEquals(long actual, int expected) {

		if (actual==expected) {
			System.out.print(success+" ");
		} else {
			System.out.println(failure);
			System.out.println(actual+"!="+expected);
			printErrors();
			status=1;
		}
	}

	protected static void assertEquals(double actual, double expected) {

		if (actual==expected) {
			System.out.print(success+" ");
		} else {
			System.out.println(failure);
			System.out.println(actual+"!="+expected);
			printErrors();
			status=1;
		}
	}

	protected static void assertEquals(boolean actual, int expected) {

		if (((actual)?1:0)==expected) {
			System.out.print(success+" ");
		} else {
			System.out.println(failure);
			System.out.println(actual+"!="+expected);
			printErrors();
			status=1;
		}
	}

	// strips a leading '$' and thousands separators, then trailing-zero
	// decimals, so money values that differ only in formatting compare equal
	private static String normalizeMoney(String value) {

		value=value.replace("$","").replace(",","");
		if (value.indexOf('.')>=0) {
			while (value.endsWith("0")) {
				value=value.substring(0,value.length()-1);
			}
			if (value.endsWith(".")) {
				value=value.substring(0,value.length()-1);
			}
		}
		return value;
	}

	// compares money values tolerantly - a leading '$', thousands separators,
	// and trailing-zero decimals are ignored, so "1.00", "$1.00", and
	// "1.0000" all match.  a real precision difference like "1.23" vs
	// "1.2345" still fails.  old freetds (0.91) renders money with 2 decimal
	// places instead of 4.
	protected static void assertMoneyEquals(String actual, String expected) {

		if (actual==null || expected==null) {
			assertEquals(actual,expected);
			return;
		}

		if (normalizeMoney(actual).equals(normalizeMoney(expected))) {
			System.out.print(success+" ");
		} else {
			System.out.println(failure);
			System.out.println(actual+"!="+expected);
			printErrors();
			status=1;
		}
	}

	// compares a rendered money length tolerantly.  old freetds (0.91)
	// renders money with 2 decimal places instead of 4, so the length may
	// be 2 short.
	protected static void assertMoneyLengthEquals(long actual, int expected) {

		if (actual==expected || (expected>=2 && actual==expected-2)) {
			System.out.print(success+" ");
		} else {
			System.out.println(failure);
			System.out.println(actual+"!="+expected);
			printErrors();
			status=1;
		}
	}

	protected static void assertTrue(boolean actual) {

		if (actual) {
			System.out.print(success+" ");
		} else {
			System.out.println(failure);
			System.out.println(actual+"!=true");
			printErrors();
			status=1;
		}
	}

	protected static void assertFalse(boolean actual) {

		if (!actual) {
			System.out.print(success+" ");
		} else {
			System.out.println(failure);
			System.out.println(actual+"!=false");
			printErrors();
			status=1;
		}
	}

	protected static void assertStartsWith(String actual, String prefix) {

		if (actual!=null && actual.startsWith(prefix)) {
			System.out.print(success+" ");
		} else {
			System.out.println(failure);
			System.out.println(actual+"!="+prefix);
			printErrors();
			status=1;
		}
	}

	protected static void assertInResultSet(SQLRCursor cursor,
						String column, String value) {

		for (long i=0; i<cursor.rowCount(); i++) {
			if (value.equals(cursor.getField(i,column))) {
				System.out.print(success+" ");
				return;
			}
		}
		System.out.println(failure);
		System.out.println("\""+value+"\" not found in column \""+
								column+"\"");
		printErrors();
		status=1;
	}

	protected static void reportTestStatus() {

		if (status==0) {
			System.out.println(alltestssucceeded);
		} else {
			System.out.println(sometestsfailed);
		}
	}
}

File: test/cs/sqlrtest.cs
// Copyright (c) David Muse
// See the file COPYING for more information.

using System;
using SQLRClient;

namespace SQLRClientTest
{
    class SQLRTest
    {
        protected static int status=0;

        protected static SQLRConnection con=null;
        protected static SQLRCursor cur=null;
        protected static SQLRConnection secondcon=null;
        protected static SQLRCursor secondcur=null;

        protected static String success="\u001B[32msuccess\u001B[0m";
        protected static String failure="\u001B[31mfailure\u001B[0m";
        protected static String alltestssucceeded="\n\u001B[34mAll tests succeeded\u001B[0m";
        protected static String sometestsfailed="\n\u001B[38;5;208mSome tests failed\u001B[0m";

        protected static void printErrors()
        {
            if (cur != null)
            {
                String err = cur.errorMessage();
                if (!String.IsNullOrEmpty(err))
                {
                    Console.WriteLine(err);
                    return;
                }
            }
            if (secondcur != null)
            {
                String err = secondcur.errorMessage();
                if (!String.IsNullOrEmpty(err))
                {
                    Console.WriteLine(err);
                    return;
                }
            }
            if (con != null)
            {
                String err = con.errorMessage();
                if (!String.IsNullOrEmpty(err))
                {
                    Console.WriteLine(err);
                    return;
                }
            }
            if (secondcon != null)
            {
                String err = secondcon.errorMessage();
                if (!String.IsNullOrEmpty(err))
                {
                    Console.WriteLine(err);
                    return;
                }
            }
        }

        protected static void pass()
        {
            Console.Write(success+" ");
            Console.Out.Flush();
        }

        protected static void fail(Object actual, Object expected)
        {
            Console.WriteLine(failure);
            Console.WriteLine("\"" + actual + "\" != \"" + expected + "\"");
            printErrors();
            Console.Out.Flush();
            status=1;
        }

        protected static void assertEquals(String actual, String expected)
        {
            if (expected == null)
            {
                if (actual == null) { pass(); } else { fail(actual,expected); }
                return;
            }
            if (actual == expected || (actual != null && actual == expected))
            {
                pass();
            }
            else if (actual != null && actual.Equals(expected))
            {
                pass();
            }
            else
            {
                fail(actual,expected);
            }
        }

        protected static void assertEquals(String actual, String expected, Int32 size)
        {
            if (expected == null)
            {
                if (actual == null) { pass(); } else { fail(actual,expected); }
                return;
            }
            if (actual != null &&
                actual.Length >= size &&
                expected.Length >= size &&
                actual.Substring(0,size) == expected.Substring(0,size))
            {
                pass();
            }
            else
            {
                fail(actual,expected);
            }
        }

        protected static void assertEquals(Byte[] actual, Byte[] expected)
        {
            if (expected == null)
            {
                if (actual == null) { pass(); } else { fail(actual,expected); }
                return;
            }
            if (actual == null || actual.Length != expected.Length)
            {
                fail(actual,expected);
                return;
            }
            for (int i = 0; i < actual.Length; i++)
            {
                if (actual[i] != expected[i])
                {
                    fail(actual,expected);
                    return;
                }
            }
            pass();
        }

        protected static void assertEquals(Byte[] actual, Byte[] expected, Int32 size)
        {
            if (expected == null)
            {
                if (actual == null) { pass(); } else { fail(actual,expected); }
                return;
            }
            if (actual == null || actual.Length < size || expected.Length < size)
            {
                fail(actual,expected);
                return;
            }
            for (int i = 0; i < size; i++)
            {
                if (actual[i] != expected[i])
                {
                    fail(actual,expected);
                    return;
                }
            }
            pass();
        }

        protected static void assertEquals(Int64 actual, Int64 expected)
        {
            if (actual == expected) { pass(); } else { fail(actual,expected); }
        }

        protected static void assertEquals(UInt64 actual, UInt64 expected)
        {
            if (actual == expected) { pass(); } else { fail(actual,expected); }
        }

        protected static void assertEquals(Int32 actual, Int32 expected)
        {
            if (actual == expected) { pass(); } else { fail(actual,expected); }
        }

        protected static void assertEquals(UInt32 actual, UInt32 expected)
        {
            if (actual == expected) { pass(); } else { fail(actual,expected); }
        }

        protected static void assertEquals(Int16 actual, Int16 expected)
        {
            if (actual == expected) { pass(); } else { fail(actual,expected); }
        }

        protected static void assertEquals(UInt16 actual, UInt16 expected)
        {
            if (actual == expected) { pass(); } else { fail(actual,expected); }
        }

        protected static void assertEquals(Double actual, Double expected)
        {
            if (actual == expected) { pass(); } else { fail(actual,expected); }
        }

        protected static void assertEquals(Boolean actual, Boolean expected)
        {
            if (actual == expected) { pass(); } else { fail(actual,expected); }
        }

        protected static void assertTrue(Boolean actual)
        {
            if (actual) { pass(); } else { fail(actual,true); }
        }

        protected static void assertFalse(Boolean actual)
        {
            if (!actual) { pass(); } else { fail(actual,false); }
        }

        protected static void assertInResultSet(SQLRCursor cursor, String column, String value)
        {
            for (UInt64 i = 0; i < cursor.rowCount(); i++)
            {
                if (cursor.getField(i, column) == value)
                {
                    pass();
                    return;
                }
            }
            Console.WriteLine(failure);
            Console.WriteLine("\"" + value + "\" not found in column \"" + column + "\"");
            printErrors();
            status = 1;
        }

        protected static void reportTestStatus()
        {
            if (status==0)
            {
                Console.WriteLine(alltestssucceeded);
            }
            else
            {
                Console.WriteLine(sometestsfailed);
            }
            Console.Out.Flush();
        }
    }
}
